festim/materials/materials.py

```diff
@@ -13,14 +14,32 @@ class Materials(list):
     """
 
     def __init__(self, *args):
-        super().__init__(*args)
+        # checks that input is list
+        try:
+            super().__init__(*args)
+        except:
+            raise TypeError("festim.Materials must be a list")
+
+        # checks that list elements are festim.Material
+        if len(self) != 0:
+            if not all(isinstance(t, festim.Material) for t in self):
+                raise TypeError("festim.Materials must be a list of festim.Material")
```

Comment:
_Here I'll make a comment on `F.Materials` assuming all modified classes_

As I understand, `.append` would also add an object of any type to `F.Traps.traps` in the previous version (Am I wrong?). In the current version, the validaty of appended items also won't be tested, yes. To add a check of item validaty to every list method, I think we have to re-define all of them (at least, those that are connected to list item assignmen: `__setitem__`, `insert`, `append`, `extend`), i.e. to make a real user-defined list. As an example, it can be done like this:
```python
class Materials(list):
    """
    A list of festim.Material objects
    """

    def __init__(self, iterable):
        # checks that input is list
        if isinstance(self, list):
            super().__init__(self._validate_material(item) for item in iterable)
        else:
            raise TypeError("festim.Materials must be a list")
        
        ...

    def __setitem__(self, index, item):
        super().__setitem__(index, self._validate_material(item))

    def insert(self, index, item):
        super().insert(index, self._validate_material(item))

    def append(self, item):
        super().append(self._validate_material(item))
        
    def _validate_material(self, value):
        if isinstance(value, festim.Material):
            return value
        raise TypeError("festim.Materials must be a list of festim.Material")
```    
@RemDelaporteMathurin what do you think?

Reply:
> As I understand, `.append` would also add an object of any type to `F.Materials.materials` in the previous version (Am I wrong?)

Yes



> As an example, it can be done like this:

This seems to be exactly what is suggested [here](https://realpython.com/inherit-python-list/)
I'm happy with it!

Later we can also refactore things a bit and do something similar like [ what is done in OpenMC](https://github.com/openmc-dev/openmc/blob/f14fc55e60463710d3f165b0a9cdc2ef9fee1534/openmc/checkvalue.py#L284) where they have a `CheckedList` class and have classes inherit from it. But let's keep this for later.

Reply:
> This seems to be exactly what is suggested [here](https://realpython.com/inherit-python-list/)

Ow, yes, but I read the russian topic on it. I suppose it was just a translation :) So, I'll give a try